docs/architecture/adr-028-public-key-addresses.md

```diff
@@ -56,113 +97,197 @@ The current multisig public keys use amino serialization to generate the address
+As in other parts of the Cosmos SDK, we will use `sha256`.
 
-We have three types of accounts we would like to create addresses for in the future:
-- regular public key addresses for new signature algorithms (ex. `sr25519`).
-- public key addresses for multisig public keys that don't use amino encoding
-- module accounts: basically any accounts which cannot sign transactions and
-which are managed internally by modules
+### Simple Account Address
 
-To address all of these use cases we propose the following basic `AddressHash` function,
-based on the discussions in [\#5694](https://github.com/cosmos/cosmos-sdk/issues/5694):
+We start with defining a hash base algorithm for generating addresses. Notably, it's used for accounts represented by a single key pair. For each public key schema we have to have an associated `typ` string, which we are discussing in a section below. `hash` is a cryptographic hash function defined in the previous section.
 
 ```go
-func AddressHash(prefix string, contents []byte) []byte {
-	preImage := []byte(prefix)
-	if len(contents) != 0 {
-		preImage = append(preImage, 0)
-		preImage = append(preImage, contents...)
-	}
-	return sha256.Sum256(preImage)[:20]
+const A_LEN = 32
+
+func Hash(typ string, key []byte) []byte {
+    return hash(hash(typ) + key)[:A_LEN]
 }
 ```
 
-`AddressHash` always take a string `prefix` as a starting point which should represent the
-type of public key (ex. `sr25519`) or module account being used (ex. `staking` or `group`).
-For public keys, the `contents` parameter is used to specify the binary contents of the public
-key. For module accounts, `contents` can be left empty (for modules which don't manage "sub-accounts"),
-or can be some module-specific content to specify different pools (ex. `bonded` or `not-bonded` for `staking`)
-or managed accounts (ex. different accounts managed by the `group` module).
+The `+` is a bytes concatenation, which doesn't use any separator.
 
-In the `preImage`, the byte value `0` is used as the separator between `prefix` and `contents`. This is a logical
-choice given that `0` is an invalid value for a string character and is commonly used as a null terminator.
+This algorithm is an outcome from a consulting session with a cryptographer.
+Motivation: this algorithm keeps the address relatively small (length of the `typ` doesn't impact on the length of the final address)
+and it's more secure than [post-hash-prefix-proposal] (which uses first 20 bytes of a pubkey hash, that significantly reduce the address space).
+Moreover the cryptographer motivated the choice to add `typ` in the hash to protect against switch table attack.
 
-### Canonical Public Key Address Prefixes
+We use `address.Hash` function for generating address for all accounts represented by a single key:
+* simple public keys: `address.Hash(keyType, pubkey)`
++ aggregated keys (eg: BLS): `address.Hash(keyType, aggregatedPubKey)`
++ module accounts: `addoress.Hash("module", moduleName)`
 
```

Comment:
```suggestion
module sub-accounts: `address.Hash("module", concat(moduleName, moduleSubKey)`
```

Reply:
This is case is discussed in the _Composed Address_ and use `Compose` function. Here I'm showing a simple module address.